---
title: "Structure function analysis of mutational effects"
author: "Tyler Starr"
date: "5/11/2020"
output:
  github_document:
  html_preview: true
editor_options: 
  chunk_output_type: inline

---
  
This notebook analyzes our single mutant effects on binding and expression in light of structural features of the RBD.

```{r setup, message=FALSE, warning=FALSE, error=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(dev.args = list(png = list(type = "cairo")))

#list of packages to install/load
packages = c("yaml","data.table","tidyverse","bio3d","gridExtra")
#install any packages not already installed
installed_packages <- packages %in% rownames(installed.packages())
if(any(installed_packages == F)){
  install.packages(packages[!installed_packages])
}
#load packages
invisible(lapply(packages, library, character.only=T))

#read in config file
config <- read_yaml("config.yaml")

#read in file giving concordance between RBD numbering and SARS-CoV-2 Spike numbering
RBD_sites <- data.table(read.csv(file="data/RBD_sites.csv",stringsAsFactors=F))

#make output directory
if(!file.exists(config$structure_function_dir)){
  dir.create(file.path(config$structure_function_dir))
}
```

Session info for reproducing environment:
```{r print_sessionInfo}
sessionInfo()
```

## Setup

Read in tables of variant effects on binding and expression, for single mutations to the SARS-CoV-2 RBD and for a panel of homolog RBDs from the sarbecovirus clade.

```{r read_data}
homologs <- data.table(read.csv(file=config$homolog_effects_file,stringsAsFactors = F))
mutants <- data.table(read.csv(file=config$single_mut_effects_file,stringsAsFactors = F))

#rename mutants site indices to prevent shared names with RBD_sites, simplifying some downstream calculations that cross-index these tables
setnames(mutants, "site_RBD", "RBD_site");setnames(mutants, "site_SARS2", "SARS2_site")

#add color column to homologs, by clade
homologs$clade_color <- as.character(NA); homologs[clade=="Clade 1",clade_color := "#EF4136"]; homologs[clade=="Clade 2",clade_color := "#009444"]; homologs[clade=="Clade 3",clade_color := "#EE2A7B"]; homologs[clade=="SARS-CoV-2",clade_color := "#2E3192"]
```

## Overall relationship between structure and function

Let's investigate how the RBD structure influences mutational effects on expression and binding. First, we compute the *mean* effect of mutations on binding and expression for each site in the structure, as well as the best (max) and worst (min) mutational effects on these two measurements (excluding nonsense and synonymous mutants).

```{r mean_max_mut_effect_per_site}
RBD_sites[,mean_bind := mean(mutants[SARS2_site==site_SARS2 & wildtype != mutant & mutant != "*",bind_avg],na.rm=T),by=site_SARS2]
RBD_sites[,max_bind := max(mutants[SARS2_site==site_SARS2 & wildtype != mutant & mutant != "*",bind_avg],na.rm=T),by=site_SARS2]
RBD_sites[,min_bind := min(mutants[SARS2_site==site_SARS2 & wildtype != mutant & mutant != "*",bind_avg],na.rm=T),by=site_SARS2]

RBD_sites[,mean_expr := mean(mutants[SARS2_site==site_SARS2 & wildtype != mutant & mutant != "*",expr_avg],na.rm=T),by=site_SARS2]
RBD_sites[,max_expr := max(mutants[SARS2_site==site_SARS2 & wildtype != mutant & mutant != "*",expr_avg],na.rm=T),by=site_SARS2]
RBD_sites[,min_expr := min(mutants[SARS2_site==site_SARS2 & wildtype != mutant & mutant != "*",expr_avg],na.rm=T),by=site_SARS2]
```

First, let's see how mutational effects on binding expression correlate at the level of individual mutations and at the level of site-level mean effects of mutation. We can see below that for many mutations and sites, mutational effects on expression and binding are intertwined. However, several positions are tolerant to mutation with respect to expression despite being sensitive to mutation with respect to binding. Below, we output the sites of these mutations, which we can visualize on the ACE2-bound RBD structure using `dms-view`, linked [here for per-site](https://dms-view.github.io/?pdb-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2F6m0j.pdb&markdown-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2FBloomLab_rbd.md&data-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2Fresults%2FBloomLab2020_rbd.csv&condition=natural+frequencies&site_metric=site_entropy&mutation_metric=mut_frequency&selected_sites=447%2C449%2C456%2C473%2C476%2C487%2C489%2C496%2C500%2C502%2C505) and [here for per-mutation](https://dms-view.github.io/?pdb-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2F6m0j.pdb&markdown-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2FBloomLab_rbd.md&data-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2Fresults%2FBloomLab2020_rbd.csv&condition=natural+frequencies&site_metric=site_entropy&mutation_metric=mut_frequency&selected_sites=443%2C455%2C456%2C475%2C487%2C489%2C496%2C498%2C500%2C501%2C502%2C505). We can see that these sites exhibiting binding-specific mutational sensitivity are invariably at the ACE2-contact interface, or in the case of at least one mutation (S443N), perhaps second shell posititions that are still ACE2-proximal.

```{r correlation_mut_expression_binding, fig.width=8,fig.height=4.25,fig.align="center", dpi=300,dev="png"}
par(mfrow=c(1,2))
plot(RBD_sites$mean_expr,RBD_sites$mean_bind,pch=19,col="#00000050",xlab="mean mutational effect on expression",ylab="mean mutational effect on binding",main="binding versus expression effects,\naverage per site")

plot(mutants$expr_avg,mutants$bind_avg,pch=19,col="#00000050",xlab="mutational effect on expression",ylab="mutational effect on binding",main="binding versus expression effects,\nper mutant")

#output sites and mutations with seemingly binding-specific detrimental effects
RBD_sites[mean_expr > -1 & mean_bind < -1,site_SARS2]
mutants[expr_avg > -0.5 & bind_avg < -2,.(SARS2_site,mutation)]
```
The relative solvent accessiibility (RSA) of an amino acid residue is known to be a dominant factor influencing its tolerance to mutation. Let's see how RSA of a position is related to its mutational sensitivity for binding. We use RSA in two different structural contexts -- the free RBD structure, and the RBD structure when complexed with ACE2. We can see that mutational sensitivity of a position with respect to binding is better described by RSA in the ACE2-bound RBD complex.

```{r mean_mut_effect_versus_RSA, fig.width=12,fig.height=4.25,fig.align="center", dpi=300,dev="png"}
par(mfrow=c(1,3))
plot(RBD_sites$RSA_unbound,RBD_sites$mean_bind,pch=19,col="gray80",xlab="Residue relative solvent accessibility, unbound",ylab="Mean effect of mutation on binding",main="Unbound RBD",xlim=c(0,1))

plot(RBD_sites$RSA_bound,RBD_sites$mean_bind,pch=19,col="gray20",xlab="Residue relative solvent accessibility, ACE2-bound",ylab="Mean effect of mutation on binding",main="ACE2-bound RBD")


plot(RBD_sites$RSA_unbound,RBD_sites$mean_bind,pch=19,col="gray80",xlab="Residue relative solvent accessibility",ylab="Mean effect of mutation on binding",main="overlaid")
points(RBD_sites$RSA_bound,RBD_sites$mean_bind,pch=19, col="gray20")

#save pdf
invisible(dev.print(pdf, paste(config$structure_function_dir,"/mean-bind_v_RSA.pdf",sep="")))

```

To further visualize site-wise mutational sensitivity on the 3D structure, let's output `.pdb` files for the ACE2-bound RBD structure in which we replace the B factor column with metrics of interest for each site: 1) the mean effect of mutation on binding, 2) the mean effect of mutation on expression, 3) the max effect of any mutation on binding, and 4) the max effect of any mutation on expression. In PyMol, we can then visualize spheres at each Calpha, colored by spectrum from low (yellow) to high (blue) for each metric by manually executing the following commands in a PyMol session in which one of the output `pdb` files is opened:
```
hide all; show cartoon
color warmpink, chain A; color gray80, chain E
set sphere_scale, 0.6
create RBD_CA, chain E and name ca
hide cartoon, RBD_CA; show spheres, RBD_CA
spectrum b, yellow green blue, RBD_CA
```

```{r output_pdbs}
pdb <- read.pdb(file="data/structures/ACE2-bound/6m0j.pdb")

#color by mean effect on binding
pdb_mean_bind <- pdb
pdb_mean_bind$atom$b <- NA
for(i in 1:nrow(pdb_mean_bind$atom)){
  res <- pdb_mean_bind$atom$resno[i]
  chain <- pdb_mean_bind$atom$chain[i]
  mean_bind <- RBD_sites[site_SARS2==res & chain_6M0J==chain & !is.na(chain_6M0J),mean_bind]
  if(length(mean_bind)>0){pdb_mean_bind$atom$b[i] <- mean_bind}else{pdb_mean_bind$atom$b[i] <- 0}
}
#save pdb
write.pdb(pdb=pdb_mean_bind,file=paste(config$structure_function_dir,"/6m0j_b-factor-mean-bind.pdb",sep=""), b = pdb_mean_bind$atom$b)

#color by max effect on binding
pdb_max_bind <- pdb
pdb_max_bind$atom$b <- NA
for(i in 1:nrow(pdb_max_bind$atom)){
  res <- pdb_max_bind$atom$resno[i]
  chain <- pdb_max_bind$atom$chain[i]
  max_bind <- RBD_sites[site_SARS2==res & chain_6M0J==chain & !is.na(chain_6M0J),max_bind]
  if(length(max_bind)>0){pdb_max_bind$atom$b[i] <- max_bind}else{pdb_max_bind$atom$b[i] <- 0}
}
#save pdb
write.pdb(pdb=pdb_max_bind,file=paste(config$structure_function_dir,"/6m0j_b-factor-max-bind.pdb",sep=""), b = pdb_max_bind$atom$b)

#color by mean effect on expression
pdb_mean_expr <- pdb
pdb_mean_expr$atom$b <- NA
for(i in 1:nrow(pdb_mean_expr$atom)){
  res <- pdb_mean_expr$atom$resno[i]
  chain <- pdb_mean_expr$atom$chain[i]
  mean_expr <- RBD_sites[site_SARS2==res & chain_6M0J==chain & !is.na(chain_6M0J),mean_expr]
  if(length(mean_expr)>0){pdb_mean_expr$atom$b[i] <- mean_expr}else{pdb_mean_expr$atom$b[i] <- 0}
}
#save pdb
write.pdb(pdb=pdb_mean_expr,file=paste(config$structure_function_dir,"/6m0j_b-factor-mean-expr.pdb",sep=""), b = pdb_mean_expr$atom$b)

#color by max effect on expression
pdb_max_expr <- pdb
pdb_max_expr$atom$b <- NA
for(i in 1:nrow(pdb_max_expr$atom)){
  res <- pdb_max_expr$atom$resno[i]
  chain <- pdb_max_expr$atom$chain[i]
  max_expr <- RBD_sites[site_SARS2==res & chain_6M0J==chain & !is.na(chain_6M0J),max_expr]
  if(length(max_expr)>0){pdb_max_expr$atom$b[i] <- max_expr}else{pdb_max_expr$atom$b[i] <- 0}
}
#save pdb
write.pdb(pdb=pdb_max_expr,file=paste(config$structure_function_dir,"/6m0j_b-factor-max-expr.pdb",sep=""), b = pdb_max_expr$atom$b)


```

## DFEs and heatmaps

Let's look at the distribution of single-mutant effects on binding, and compare the fraction of mutations that are within the window defined by known functional RBD homologs for these two phenotypes.

For the binding plot on the left, the intermediate blue point on the x-scale is RaTG13, which *can* promote huACE2-mediated cell entry in in vitro cellular infection assays (though less efficiently than SARS-CoV-2), though whether this is sufficient to enable efficient viral replication in more complex models is uncertain. For the cluster of points near 0, the farthest-left point is LYRa11, which according to Letko et al. can also promote huACE2-mediated cellular entry, though less efficiently than SARS-CoV-1 and other bat CoV isolates such as WIV1/16 (identical RBDs). Therefore, these two points define a window of affinites that can at least support in vitro cellular infection -- but in reality, the window of possible "neutrality" with regards to actual human infectivity is probably better set by the remaining four points with delta log<sub>10</sub>(*K*<sub>A,app</sub>) values ~ 0, consisting of SARS-CoV-1, WIV1/16, SARS-CoV-2, and GD-Pangolin RBDs, in that rank-order. Taken together, we identify `r nrow(mutants[wildtype != mutant & !is.na(bind_avg) & bind_avg >= homologs[homolog=="SARS-CoV-1",bind_avg],])` single mutants (`r round(nrow(mutants[wildtype != mutant & !is.na(bind_avg) & bind_avg >= homologs[homolog=="SARS-CoV-1",bind_avg],])/nrow(mutants[wildtype != mutant & !is.na(bind_avg),])*100,digits=2)`%) whose affinity effects are mild enough to potentially enable human infectivity (SARS-CoV-1 cutoff), and `r nrow(mutants[wildtype != mutant & !is.na(bind_avg) & bind_avg >= homologs[homolog=="RaTG13",bind_avg],])` single mutants (`r round(nrow(mutants[wildtype != mutant & !is.na(bind_avg) & bind_avg >= homologs[homolog=="RaTG13",bind_avg],])/nrow(mutants[wildtype != mutant & !is.na(bind_avg),])*100,digits=2)`%) whose affinity is potentially sufficient to enable in vitro cellular infectivity (RaTG13 cutoff). Taken together, this suggests a quite large sequence space of RBD diversity that is consistent with huACE2 binding and entry.

```{r DFE_bind, fig.width=4.25,fig.height=4.25,fig.align="center", dpi=300,dev="png"}
plot(density(mutants[wildtype != mutant,bind_avg],na.rm=T,adjust=0.5),main="SARS-CoV-2 mutant log10Ka,app",xlab="delta log10(KA,app)"); polygon(density(mutants[wildtype != mutant,bind_avg],na.rm=T,adjust=0.5),col="gray50",border="black")
set.seed(100);points(homologs[,bind_avg],sample(seq(1.45,1.55,by=0.005),nrow(homologs)),col=homologs[,clade_color],pch=19,cex=1.1)
# plot(density(mutants[wildtype != mutant & mutant != "*",expr_avg],na.rm=T,adjust=0.5),main="SARS-CoV-2 mutant expression",xlab="delta expression"); polygon(density(mutants[wildtype != mutant & mutant != "*",expr_avg],na.rm=T,adjust=0.5),col="gray50",border="black")
# set.seed(100);points(homologs[,expr_avg],sample(seq(0.63,0.67,by=0.002),nrow(homologs)),col=homologs[,clade_color],pch=19,cex=1.1)
#can also make plot for expression, but our expression range is less well calibrated, probably because of the issue with rare barcodes exhibiting lower than expected expression probably due to unseen defects in the plasmid.
```

Next, let's make heatmaps of per-amino acid mutational effects on binding and expression. We first make these heatmaps for all mutations at all sites, colored by delta log<sub>10</sub>(*K*<sub>A,app</sub>). (I would eventually like to flesh out these heatmap by providing additional indicator variables as 'heatmap' style rows on the top -- things could include a color scale for RSA, conservation, an asterisk or something to indicate contact residues, indicators for NLGS or disulfides, an indicator for strucutrla contacts in full length Spike trimer, etc. I also want to think about whether there are better color scales to use, including the "divergence" in scale between the blue (goes from white to bluest in a 0.5-unit scale), versus red (0 to reddest in a 5-unit scale). Happy for additional things to think about to refine these plots):
  
```{r heatmap_binding_all, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=FALSE}
#order mutant as a factor for grouping by rough biochemical grouping
mutants$mutant <- factor(mutants$mutant, levels=c("*","C","P","G","V","M","L","I","A","F","W","Y","T","S","N","Q","E","D","H","K","R"))
#add character vector indicating wildtype to use as plotting symbols for wt
mutants[,wildtype_indicator := ""]
mutants[mutant==wildtype,wildtype_indicator := "x"]

p1 <- ggplot(mutants[mutant!="*" & SARS2_site %in% seq(331,431),],aes(SARS2_site,mutant))+geom_tile(aes(fill=bind_avg))+
  scale_fill_gradientn(colours=c("#A94E35","#F48365","#FFFFFF","#7378B9","#383C6C"),limits=c(-5,0.5),values=c(0,2.5/5.5,5/5.5,5.25/5.5,5.5/5.5),na.value="gray")+
  scale_x_continuous(expand=c(0,0),breaks=c(331,seq(340,430,by=5)))+
  labs(x="RBD site",y="mutant")+theme_classic(base_size=9)+
  coord_equal()+
  geom_text(aes(label=wildtype_indicator),size=2,color="gray10")

p2 <- ggplot(mutants[mutant!="*" & SARS2_site %in% seq(432,531),],aes(SARS2_site,mutant))+geom_tile(aes(fill=bind_avg))+
  scale_fill_gradientn(colours=c("#A94E35","#F48365","#FFFFFF","#7378B9","#383C6C"),limits=c(-5,0.5),values=c(0,2.5/5.5,5/5.5,5.25/5.5,5.5/5.5),na.value="gray")+
  scale_x_continuous(expand=c(0,0),breaks=c(432,seq(440,530,by=5)))+
  labs(x="RBD site",y="mutant")+theme_classic(base_size=9)+
  coord_equal()+
  geom_text(aes(label=wildtype_indicator),size=2,color="gray10")

grid.arrange(p1,p2,ncol=1)
```
And next, the same heat map, colored by delta mean fluorescence (expression) (once again, happy for thoughts. Do we like having the stop variants in here (and altering the scale?):

```{r heatmap_expression_all, fig.width=12,fig.height=6,fig.align="center", dpi=500,dev="png",echo=FALSE}
p1 <- ggplot(mutants[SARS2_site %in% seq(331,431),],aes(SARS2_site,mutant))+geom_tile(aes(fill=expr_avg))+
  scale_fill_gradientn(colours=c("#A94E35","#F48365","#FFFFFF","#7378B9","#383C6C"),limits=c(-5,1),values=c(0,2.5/6,5/6,5.5/6,6/6),na.value="gray")+
  scale_x_continuous(expand=c(0,0),breaks=c(331,seq(340,430,by=5)))+
  labs(x="RBD site",y="mutant")+theme_classic(base_size=9)+
  coord_equal()+
  geom_text(aes(label=wildtype_indicator),size=2,color="gray10")

p2 <- ggplot(mutants[SARS2_site %in% seq(432,531),],aes(SARS2_site,mutant))+geom_tile(aes(fill=expr_avg))+
  scale_fill_gradientn(colours=c("#A94E35","#F48365","#FFFFFF","#7378B9","#383C6C"),limits=c(-5,1),values=c(0,2.5/6,5/6,5.5/6,6/6),na.value="gray")+
  scale_x_continuous(expand=c(0,0),breaks=c(432,seq(440,530,by=5)))+
  labs(x="RBD site",y="mutant")+theme_classic(base_size=9)+
  coord_equal()+
  geom_text(aes(label=wildtype_indicator),size=2,color="gray10")

grid.arrange(p1,p2,ncol=1)
```

We next make these heatmaps, zooming in on residues in special structural and functional classes. (I collapse the heatmap color range for all expression values below 4, so the range is better calibrated between expression and binding -- the range of missense variants for expression goes to 4.07, only nonsense mutants span to 5). First, let's visualize heatmaps of paired cysteines that form disulfides. The following heatmaps reorder sites by cysteine pair (I would love to add a line at the top grouping the paired cysteines, or create a small gap between each pair of columns, but I am gg-inept and so this will do for now)

We can see that cysteines within a disulfide pair have similiar sensitivities to mutation and even similar biochemical preferences, but there are differing patterns of sensitivity for binding and expression. Disulfide pair 4 is the most sensitive to mutation with respect to binding, whereas it only is moderately important for expression -- this is the disulfide pair within the RBM loop, that stabilizes regions of the ACE2 interface. The remaining three disulfides are in the core RBD domain, and show similar sensitivities for expression and binding -- pair 3 is tolerant to mutation, pair 1 is moderately sensitive, and pair 2 is most sensitive -- though hydrophobic mutations have noticeably decreased deleterious effect for binding, which is cool to see.

This trend is consistent with a series of Cys->Ala mutations made in SARS-CoV-2 RBD in [Wong et al. JBC 2004](https://www.jbc.org/content/279/5/3197.short) (I report the numbers converted to SARS-CoV-2 numbering): mutations at SARS-CoV-1 RBD positions 379 and 432 severely impaired expression, mutations at 361, 480, and 488 had only mild effects on expression but strongly decreased ACE2-binding, and mutations at 336 and 391 had little effect on expression or binding (this is in 331-524 RBD construct, so lacking cysteine 525).

```{r heatmap_bind_expr_disulfide, fig.width=4,fig.height=4,fig.align="center", dpi=500,dev="png",echo=FALSE}
muts_temp <- mutants[SARS2_site %in% RBD_sites[!is.na(disulfide),site_SARS2],];muts_temp$SARS2_site <- as.factor(muts_temp$SARS2_site)
muts_temp[,disulfide:=RBD_sites[site_SARS2==SARS2_site,disulfide],by=mutation];muts_temp$disulfide <- as.factor(muts_temp$disulfide)


p1 <- ggplot(muts_temp,aes(paste(disulfide,SARS2_site),mutant))+geom_tile(aes(fill=bind_avg))+
  scale_fill_gradientn(colours=c("#A94E35","#F48365","#FFFFFF","#7378B9","#383C6C"),limits=c(-5,0.5),values=c(0,2.5/5.5,5/5.5,5.25/5.5,5.5/5.5),na.value="gray")+
  scale_x_discrete(expand=c(0,0),labels=as.character(unique(muts_temp[order(disulfide,SARS2_site),SARS2_site])))+
  labs(x="RBD site, grouped\nby disulfide pair",y="mutant")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x = element_text(angle=45,hjust=1))+
  geom_text(aes(label=wildtype_indicator),size=2,color="gray10")

p2 <- ggplot(muts_temp,aes(paste(disulfide,SARS2_site),mutant))+geom_tile(aes(fill=expr_avg))+
  scale_fill_gradientn(colours=c("#A94E35","#A94E35","#F48365","#FFFFFF","#7378B9","#383C6C"),limits=c(-5,1),values=c(0,1/6,3/6,5/6,5.5/6,6/6),na.value="gray")+
  scale_x_discrete(expand=c(0,0),labels=as.character(unique(muts_temp[order(disulfide,SARS2_site),SARS2_site])))+
  labs(x="RBD site, grouped\nby disulfide pair",y="mutant")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x = element_text(angle=45,hjust=1))+
  geom_text(aes(label=wildtype_indicator),size=2,color="gray10")

grid.arrange(p1,p2,ncol=2)

```

Next, let's look at the N-linked glycosylation sites. In particular, we look at mutational effects at N331 and T333, N343 and T345, as well as N370 and A372, which in SARS-CoV-1 is an NST motif NLGS. A prior paper, [Chen et al. 2014](https://www.tandfonline.com/doi/full/10.4161/hv.27464), showed that in SARS-CoV-1 RBD produced in Pichia yeast, yields were lowered when progressively knocking out each of these glycans. (They didn't do individual knockouts, it seems.) We also look at the effects of mutations i+2 from surface asparagines, where mutations to S or T could introduce new NLGS.

The two NLGS in the SARS-CoV-2 RBD are important for stability. We can see similar effects of mutating the focal asparagine or the +2 residue (except mutations from T to S are well-tolerated), both of which point to the N343 glycan being more important for stability. Re-introducing the SARS-CoV-1 NLGS has just a mild deleterious effect on expression. None of these three glycan mutants have major impacts on binding, consistent with their distance from the ACE2-interface.

At other asparagines, I don't see many strong indications of major effects for introducing possible glycans with +2 residue mutations to S/T. The only sites that possibly stick out, are that adding an NLGS to N501 with 503 T/S mutations has a mild deleterious effect on binding (consistent with this residue [being at the ACE2-interface](https://dms-view.github.io/?pdb-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2F6m0j.pdb&markdown-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2FBloomLab_rbd.md&data-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2Fresults%2FBloomLab2020_rbd.csv&condition=natural+frequencies&site_metric=site_entropy&mutation_metric=mut_frequency&selected_sites=501%2C503)), adding an NLGS to N487 (also an interface residue) may also be especially detrimental, but it's hard to say since many mutations besides S/T are very deleterious at position 489. More broadly, it seems there are quite a few positions where introducing NxS/T NLGS motifs is not detrimental for expression, suggesting NLGS modifications could be used to mask or resurface engineered RBD constructs for vaccine design or epitope-specific probes for identifying neutralizing antibodies. The asparagines at positions 354, 360, 388, 394, 437, 439, 448, 450, 460, 481, 487, and 501 may all be (at least individually) tolerated with regards to RBD stability -- these sites are [visualized on the RBD structure](https://dms-view.github.io/?pdb-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2F6m0j.pdb&markdown-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2FBloomLab_rbd.md&data-url=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-view%2FSARS-CoV-2%2Fmaster%2Fdata%2FSpike%2FBloomLab2020%2Fresults%2FBloomLab2020_rbd.csv&condition=natural+frequencies&site_metric=site_entropy&mutation_metric=mut_frequency&selected_sites=354%2C360%2C388%2C394%2C437%2C439%2C448%2C450%2C460%2C481%2C487%2C501), illustrating that many surfaces could be masked (though not all), and that even with the RBM, where many ACE2-competitive neutralizing Abs bind, one could engineer in glycans that block one half of the interface or the other to target certain potential epitopes. (You can also see that some of these asparagines do make structural contacts within little loops, particularly around the RBM, so they might not be exposed enough to be actually glycosylated even with the introduction of the +2 S/T)

```{r heatmap_bind_expr_NLGS, fig.width=12,fig.height=4,fig.align="center", dpi=500,dev="png",echo=FALSE}
muts_temp <- mutants[SARS2_site %in% RBD_sites[!is.na(NLGS) & NLGS != "surfaceN",site_SARS2],];muts_temp$SARS2_site <- as.factor(muts_temp$SARS2_site)
muts_temp[,NLGS:=RBD_sites[site_SARS2==SARS2_site,NLGS],by=mutation];muts_temp$NLGS <- factor(muts_temp$NLGS, levels=c("putative1","putative2","SARS-CoV-1"))


p1 <- ggplot(muts_temp,aes(paste(NLGS,SARS2_site),mutant))+geom_tile(aes(fill=bind_avg))+
  scale_fill_gradientn(colours=c("#A94E35","#F48365","#FFFFFF","#7378B9","#383C6C"),limits=c(-5,0.5),values=c(0,2.5/5.5,5/5.5,5.25/5.5,5.5/5.5),na.value="gray")+
  scale_x_discrete(expand=c(0,0),labels=as.character(unique(muts_temp[order(NLGS,SARS2_site),SARS2_site])))+
  labs(x="RBD site, grouped\nby NLGS motif",y="mutant")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x = element_text(angle=45,hjust=1))+
  geom_text(aes(label=wildtype_indicator),size=2,color="gray10")

p2 <- ggplot(muts_temp,aes(paste(NLGS,SARS2_site),mutant))+geom_tile(aes(fill=expr_avg))+
  scale_fill_gradientn(colours=c("#A94E35","#A94E35","#F48365","#FFFFFF","#7378B9","#383C6C"),limits=c(-5,1),values=c(0,1/6,3/6,5/6,5.5/6,6/6),na.value="gray")+
  scale_x_discrete(expand=c(0,0),labels=as.character(unique(muts_temp[order(NLGS,SARS2_site),SARS2_site])))+
  labs(x="RBD site, grouped\nby NLGS motif",y="mutant")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x = element_text(angle=45,hjust=1))+
  geom_text(aes(label=wildtype_indicator),size=2,color="gray10")

muts_temp <- mutants[SARS2_site %in% c(RBD_sites[NLGS == "surfaceN",site_SARS2]+2),];muts_temp$SARS2_site <- factor(muts_temp$SARS2_site)

p3 <- ggplot(muts_temp,aes(SARS2_site,mutant))+geom_tile(aes(fill=bind_avg))+
  scale_fill_gradientn(colours=c("#A94E35","#F48365","#FFFFFF","#7378B9","#383C6C"),limits=c(-5,0.5),values=c(0,2.5/5.5,5/5.5,5.25/5.5,5.5/5.5),na.value="gray")+
  scale_x_discrete(expand=c(0,0),labels=as.character(unique(muts_temp$SARS2_site)))+
  labs(x="RBD site,\nN+2 residues",y="mutant")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x = element_text(angle=45,hjust=1))+
  geom_text(aes(label=wildtype_indicator),size=2,color="gray10")

p4 <- ggplot(muts_temp,aes(SARS2_site,mutant))+geom_tile(aes(fill=expr_avg))+
  scale_fill_gradientn(colours=c("#A94E35","#A94E35","#F48365","#FFFFFF","#7378B9","#383C6C"),limits=c(-5,1),values=c(0,1/6,3/6,5/6,5.5/6,6/6),na.value="gray")+
  scale_x_discrete(expand=c(0,0),labels=as.character(unique(muts_temp$SARS2_site)))+
  labs(x="RBD site, grouped\nby NLGS motif",y="mutant")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x = element_text(angle=45,hjust=1))+
  geom_text(aes(label=wildtype_indicator),size=2,color="gray10")


grid.arrange(p1,p2,p3,p4,nrow=1,widths=c(1,1,1.8,1.8))

```
 
(next up, zoom in on residues in the RBM, contact residues, "key" contacts [sites of adaptation within SARS-CoV-1], sites with diversity in SARS-CoV-2 clade, differences between SARS-CoV-1 and SARS-CoV-2)


Does mutational tolerance with respect to binding and/or expression differ systematically between positions in the core-RBD versus the RBM loops?
```{r }

```



Next let's look at beneficial mutations on binding and expression. We want to look at where they occur, and identify any interesting hypothesized biochemical mechanisms, etc. Pick some to validate!


```{r }

```

## Output for dms-view visualization of mutational effects



